numcodecs/pcodec.py

```diff
@@ -0,0 +1,78 @@
+from typing import Optional, Literal
+
+
+import pcodec
+
+
+import numcodecs
+import numcodecs.abc
+from numcodecs.compat import ensure_contiguous_ndarray
+
+
+class PCodec(numcodecs.abc.Codec):
+    """
+    PCodec (or pco, pronounced "pico") losslessly compresses and decompresses
+    numerical sequences with high compression ratio and fast speed.
+
+    See `PCodec Repo <https://github.com/mwlon/pcodec>`_ for more information.
+
+    PCodec supports only the following numerical dtypes: uint32, unit64, int32,
+    int64, float32, and float64.
+
+    Parameters
+    ----------
+    level : int
+        A compression level from 0-12, where 12 take the longest and compresses
+        the most.
+    delta_encoding_order : init or None
+        Either a delta encoding level from 0-7 or None. If set to None, pcodec
+        will try to infer the optimal delta encoding order.
+    int_mult_spec : {'enabled', 'disabled'}
+        If enabled, pcodec will consider using int mult mode, which can
+        substantially improve compression ratio but decrease speed in some cases
+        for integer types.
+    float_mult_spec : {'enabled', 'disabled'}
+        If enabled, pcodec will consider using float mult mode, which can
+        substantially improve compression ratio but decrease speed in some cases
+        for float types.
+    param max_page_n : int
+        The maximum number of values to encoder per pcodec page.
+        If set too high or too low, pcodec's compression ratio may drop.
+    """
+
+    codec_id = "pcodec"
+
+    def __init__(
+        self,
+        level: int = 8,
+        delta_encoding_order: Optional[int] = None,
+        int_mult_spec: Literal["enabled", "disabled"] = "enabled",
+        float_mult_spec: Literal["enabled", "disabled"] = "enabled",
+        max_page_n: int = 262144,
+    ):
+        # note that we use `level` instead of `compression_level` to
+        # match other codecs
+        self.level = level
+        self.delta_encoding_order = delta_encoding_order
+        self.int_mult_spec = int_mult_spec
+        self.float_mult_spec = float_mult_spec
+        self.max_page_n = max_page_n
+
+    def encode(self, buf):
```

Comment:
```suggestion
    def encode(self, buf) -> bytes:
```
?
(This may be a good way to label array-bytes codecs; maybe also type for `buf` should be `ndarray`)

Reply:
I'm not convinced that there is much value in adding these sorts of type hints if we are not actually running type checking on the library.

Reply:
Is it not for other users of the library, and their IDEs?

Reply:
But incorrect type hints are worse than none at all! For example, is `ndarray` really the correct type for buf? Maybe, but who knows? I could add it, but without running mypy, we'll never know for sure.

Reply:
True that correctness is important, of course; but this is like the "light" version of array->array V array->bytes V bytes->bytes. Still useful. You can always get around mypy too, if you want.